File: py/src/tablut/ashton/capturerules.py
from tablut import TablutState, Action, Player, Pawn, Coord
from tablut.action import Coords


class AshtonCaptureRules:
    @staticmethod
    def remove_captured(state: TablutState, action: Action):
        pawn_capture_function = None

        if state.turn == Player.BLACK:
            AshtonCaptureRules._remove_captured_king_from_black(state, action)
            pawn_capture_function = AshtonCaptureRules._remove_captured_white_from_black

        elif state.turn == Player.WHITE:
            pawn_capture_function = AshtonCaptureRules._remove_captured_black_from_white

        AshtonCaptureRules._remove_captured_simple_pawns_from(state, action, pawn_capture_function)

        # state.moves_without_capturing += 1

    @staticmethod
    def _remove_captured_simple_pawns_from(state: TablutState, action: Action, remove_simple_pawn):
        # capture on the right
        if action.to.column < state.board.shape[1] - 2:
            captured = Coord(action.to.row, action.to.column + 1)
            second = Coord(action.to.row, action.to.column + 2)
            remove_simple_pawn(state, captured, second)

        # capture on the left
        if action.to.column > 1:
            captured = Coord(action.to.row, action.to.column - 1)
            second = Coord(action.to.row, action.to.column - 2)
            remove_simple_pawn(state, captured, second)

        # capture on the top
        if action.to.row > 1:
            captured = Coord(action.to.row - 1, action.to.column)
            second = Coord(action.to.row - 2, action.to.column)
            remove_simple_pawn(state, captured, second)

        # capture on the bottom
        if action.to.row < state.board.shape[0] - 2:
            captured = Coord(action.to.row + 1, action.to.column)
            second = Coord(action.to.row + 2, action.to.column)
            remove_simple_pawn(state, captured, second)

    @staticmethod
    def _remove_captured_black_from_white(state: TablutState, captured: Coord, second: Coord):
        """Remove black pawn between first and second white pawn"""
        captured_pawn = state.pawn(captured)
        second_pawn = state.pawn(second)

        if captured_pawn == Pawn.BLACK:
            if Pawn.is_white(second_pawn) or second_pawn == Pawn.THRONE \
                    or (second in state.board.citadels and second not in [Coords.I5, Coords.E1, Coords.A5, Coords.E9]):
                state.remove_pawn(captured)
                # state.moves_without_capturing = -1

    @staticmethod
    def _remove_captured_white_from_black(state: TablutState, captured: Coord, second: Coord):
        """Remove white pawn between first and second black pawn"""
        captured_pawn = state.pawn(captured)
        second_pawn = state.pawn(second)

        if captured_pawn == Pawn.WHITE and \
                (second_pawn == Pawn.BLACK or second_pawn == Pawn.THRONE
                 or second in state.board.citadels or second == Coords.E5):
            state.remove_pawn(captured)
            # state.moves_without_capturing = -1

    @staticmethod
    def _remove_captured_king_from_black(state: TablutState, action: Action):
        right = Coord(action.to.row, action.to.column + 1)
        left = Coord(action.to.row, action.to.column - 1)
        bottom = Coord(action.to.row + 1, action.to.column)
        top = Coord(action.to.row - 1, action.to.column)

        # capture on the right
        if action.to.column < state.board.shape[1] - 2 and state.pawn(right) == Pawn.KING:
            # re sul trono
            if right == Coords.E5:
                if state.pawn(Coords.E4) == Pawn.BLACK and state.pawn(Coords.F5) == Pawn.BLACK and state.pawn(
                        Coords.E6) == Pawn.BLACK:
                    state.remove_pawn(right)

            # re adiacente al trono
            elif right == Coords.E4:
                if state.pawn(Coords.E3) == Pawn.BLACK and state.pawn(Coords.F4) == Pawn.BLACK:
                    state.remove_pawn(right)

            elif right == Coords.E6:
                if state.pawn(Coords.F6) == Pawn.BLACK and state.pawn(Coords.E7) == Pawn.BLACK:
                    state.remove_pawn(right)

            elif right == Coords.D5:
                if state.pawn(Coords.D4) == Pawn.BLACK and state.pawn(Coords.D6) == Pawn.BLACK:
                    state.remove_pawn(right)

            # sono fuori dalle zone del trono
            else:
                if state.pawn(Coord(action.to.row, action.to.column + 2)) == Pawn.BLACK \
                        or Coord(action.to.row, action.to.column + 2) in state.board.citadels:
                    state.remove_pawn(right)

        # capture on the left
        elif action.to.column > 1 and state.pawn(left) == Pawn.KING:
            # re sul trono
            if left == Coords.E5:
                if state.pawn(Coords.E4) == Pawn.BLACK and state.pawn(Coords.D5) == Pawn.BLACK and state.pawn(
                        Coords.E6) == Pawn.BLACK:
                    state.remove_pawn(left)

            # re adiacente al trono
            elif left == Coords.E4:
                if state.pawn(Coords.E3) == Pawn.BLACK and state.pawn(Coords.D4) == Pawn.BLACK:
                    state.remove_pawn(left)

            elif left == Coords.F5:
                if state.pawn(Coords.F6) == Pawn.BLACK and state.pawn(Coords.F4) == Pawn.BLACK:
                    state.remove_pawn(left)

            elif left == Coords.E6:
                if state.pawn(Coords.E7) == Pawn.BLACK and state.pawn(Coords.D6) == Pawn.BLACK:
                    state.remove_pawn(left)

            # sono fuori dalle zone del trono
            else:
                if state.pawn(Coord(action.to.row, action.to.column - 2)) == Pawn.BLACK \
                        or Coord(action.to.row, action.to.column - 2) in state.board.citadels:
                    state.remove_pawn(left)

        # capture on the bottom
        elif action.to.row < state.board.shape[0] - 2 and state.pawn(bottom) == Pawn.KING:
            # re sul trono
            if bottom == Coords.E5:
                if state.pawn(Coords.E6) == Pawn.BLACK and state.pawn(Coords.F5) == Pawn.BLACK and state.pawn(
                        Coords.D5) == Pawn.BLACK:
                    state.remove_pawn(bottom)

            # re adiacente al trono
            elif bottom == Coords.E4:
                if state.pawn(Coords.D4) == Pawn.BLACK and state.pawn(Coords.F4) == Pawn.BLACK:
                    state.remove_pawn(bottom)

            elif bottom == Coords.D5:
                if state.pawn(Coords.C5) == Pawn.BLACK and state.pawn(Coords.D6) == Pawn.BLACK:
                    state.remove_pawn(bottom)

            elif bottom == Coords.F5:
                if state.pawn(Coords.G5) == Pawn.BLACK and state.pawn(Coords.F6) == Pawn.BLACK:
                    state.remove_pawn(bottom)

            # sono fuori dalle zone del trono
            else:
                if state.pawn(Coord(action.to.row + 2, action.to.column)) == Pawn.BLACK \
                        or Coord(action.to.row + 2, action.to.column) in state.board.citadels:
                    state.remove_pawn(bottom)

        # capture on top
        if action.to.row > 1 and state.pawn(top) == Pawn.KING:
            # re sul trono
            if top == Coords.E5:
                if state.pawn(Coords.E4) == Pawn.BLACK and state.pawn(Coords.F5) == Pawn.BLACK and state.pawn(
                        Coords.D5) == Pawn.BLACK:
                    state.remove_pawn(top)

            # re adiacente al trono
            elif top == Coords.E6:
                if state.pawn(Coords.D6) == Pawn.BLACK and state.pawn(Coords.F6) == Pawn.BLACK:
                    state.remove_pawn(top)

            elif top == Coords.D5:
                if state.pawn(Coords.C5) == Pawn.BLACK and state.pawn(Coords.D4) == Pawn.BLACK:
                    state.remove_pawn(top)

            elif top == Coords.F5:
                if state.pawn(Coords.G5) == Pawn.BLACK and state.pawn(Coords.F4) == Pawn.BLACK:
                    state.remove_pawn(top)

            # sono fuori dalle zone del trono
            else:
                if state.pawn(Coord(action.to.row - 2, action.to.column)) == Pawn.BLACK \
                        or Coord(action.to.row - 2, action.to.column) in state.board.citadels:
                    state.remove_pawn(top)

File: java/src/main/java/it/ai/game/tablut/ashton/CaptureRules.java
package it.ai.game.tablut.ashton;

import com.google.common.collect.Sets;
import it.ai.game.tablut.*;

import java.util.Set;

final class CaptureRules {
    @FunctionalInterface
    interface TriConsumer<A, B, C> {
        void apply(A a, B b, C c);
    }

    public static void removeCapturedPawns(TablutState state, Action action) {
        TriConsumer<TablutState, Coords, Coords> simplePawnCapturer = null;

        if (state.getTurn() == Player.BLACK) {
            removeKingIfCaptured(state, action);
            simplePawnCapturer = CaptureRules::removeWhitePawnIfCaptured;
        } else if (state.getTurn() == Player.WHITE) {
            simplePawnCapturer = CaptureRules::removeBlackPawnIfCaptured;
        }

        removeCapturedNormalPawns(state, action, simplePawnCapturer);
    }

    private static void removeCapturedNormalPawns(TablutState state, Action action,
                                                  TriConsumer<TablutState, Coords, Coords> removeSimplePawn) {

        int toRow = action.getTo().getRow();
        int toColumn = action.getTo().getColumn();

        // capture on the right
        if (toColumn < state.getBoard().numberOfColumns() - 2) {
            Coords captured = new Coords(toRow, toColumn + 1);
            Coords second = new Coords(toRow, toColumn + 2);
            removeSimplePawn.apply(state, captured, second);
        }

        //capture on the left
        if (toColumn > 1) {
            Coords captured = new Coords(toRow, toColumn - 1);
            Coords second = new Coords(toRow, toColumn - 2);
            removeSimplePawn.apply(state, captured, second);
        }

        //capture on the top
        if (toRow > 1) {
            Coords captured = new Coords(toRow - 1, toColumn);
            Coords second = new Coords(toRow - 2, toColumn);
            removeSimplePawn.apply(state, captured, second);
        }

        //capture on the bottom
        if (toRow < state.getBoard().numberOfRows() - 2) {
            Coords captured = new Coords(toRow + 1, toColumn);
            Coords second = new Coords(toRow + 2, toColumn);
            removeSimplePawn.apply(state, captured, second);
        }
    }

    private static final Set<Coords> whiteUnusableCitadels = Sets.newHashSet(
            Coords.I5, Coords.E1, Coords.A5, Coords.E9);

    /***
     * Remove black pawn between first and second white pawn.
     */
    private static void removeBlackPawnIfCaptured(TablutState state, Coords captured, Coords second) {
        Board board = state.getBoard();
        int captured_pawn = board.get(captured);
        int secondPawn = board.get(second);

        if (captured_pawn == Pawn.BLACK) {
            boolean pawnIsWhiteOrThrone = Pawn.isWhite(secondPawn) || second.equals(Coords.E5);

            // if second pawn is white or throne or is a valid citadel
            if (pawnIsWhiteOrThrone || (board.inCitadels(second) && !whiteUnusableCitadels.contains(second)))
                board.remove(captured);
        }
    }

    /***
     * Remove white pawn between first and second black pawn
     */
    private static void removeWhitePawnIfCaptured(TablutState state, Coords captured, Coords second) {
        Board board = state.getBoard();
        int captured_pawn = board.get(captured);
        int secondPawn = board.get(second);

        if (captured_pawn == Pawn.WHITE) {
            if (secondPawn == Pawn.BLACK || board.inCitadels(second) || second.equals(Coords.E5)) {
                board.remove(captured);
            }
        }
    }

    private static void removeKingIfCaptured(TablutState state, Action action) {
        int toRow = action.getTo().getRow();
        int toColumn = action.getTo().getColumn();
        Board board = state.getBoard();

        Coords right = new Coords(toRow, toColumn + 1);
        Coords left = new Coords(toRow, toColumn - 1);
        Coords bottom = new Coords(toRow + 1, toColumn);
        Coords top = new Coords(toRow - 1, toColumn);

        //capture on the right
        if (toColumn < board.numberOfColumns() - 2 && board.get(right) == Pawn.KING) {
            //re sul trono
            if (right.equals(Coords.E5)) {
                if (board.get(Coords.E4) == Pawn.BLACK
                        && board.get(Coords.F5) == Pawn.BLACK
                        && board.get(Coords.E6) == Pawn.BLACK) {
                    board.remove(right);
                }
            }

            //re adiacente al trono
            else if (right.equals(Coords.E4)) {
                if (board.get(Coords.E3) == Pawn.BLACK && board.get(Coords.F4) == Pawn.BLACK)
                    board.remove(right);
            } else if (right.equals(Coords.E6)) {
                if (board.get(Coords.F6) == Pawn.BLACK && board.get(Coords.E7) == Pawn.BLACK)
                    board.remove(right);
            } else if (right.equals(Coords.D5)) {
                if (board.get(Coords.D4) == Pawn.BLACK && board.get(Coords.D6) == Pawn.BLACK)
                    board.remove(right);
            }

            //sono fuori dalle zone del trono
            else {
                Coords secondPawn = new Coords(toRow, toColumn + 2);
                if (board.get(secondPawn) == Pawn.BLACK || board.inCitadels(secondPawn))
                    board.remove(right);
            }
        }

        //capture on the left
        else if (toColumn > 1 && board.get(left) == Pawn.KING) {
            //re sul trono
            if (left.equals(Coords.E5)) {
                if (board.get(Coords.E4) == Pawn.BLACK
                        && board.get(Coords.D5) == Pawn.BLACK
                        && board.get(Coords.E6) == Pawn.BLACK)
                    board.remove(left);
            }
            //re adiacente al trono
            else if (left.equals(Coords.E4)) {
                if (board.get(Coords.E3) == Pawn.BLACK && board.get(Coords.D4) == Pawn.BLACK)
                    board.remove(left);
            } else if (left.equals(Coords.F5)) {
                if (board.get(Coords.F6) == Pawn.BLACK && board.get(Coords.F4) == Pawn.BLACK)
                    board.remove(left);
            } else if (left.equals(Coords.E6)) {
                if (board.get(Coords.E7) == Pawn.BLACK && board.get(Coords.D6) == Pawn.BLACK)
                    board.remove(left);
            }
            //sono fuori dalle zone del trono
            else {
                Coords secondPawn = new Coords(toRow, toColumn - 2);
                if (board.get(secondPawn) == Pawn.BLACK || board.inCitadels(secondPawn))
                    board.remove(left);
            }
        }

        //capture on the bottom
        else if (toRow < board.numberOfRows() - 2 && board.get(bottom) == Pawn.KING) {
            //re sul trono
            if (bottom.equals(Coords.E5)) {
                if (board.get(Coords.E6) == Pawn.BLACK
                        && board.get(Coords.F5) == Pawn.BLACK
                        && board.get(Coords.D5) == Pawn.BLACK)
                    board.remove(bottom);
            }

            //re adiacente al trono
            else if (bottom.equals(Coords.E4)) {
                if (board.get(Coords.D4) == Pawn.BLACK && board.get(Coords.F4) == Pawn.BLACK)
                    board.remove(bottom);
            } else if (bottom.equals(Coords.D5)) {
                if (board.get(Coords.C5) == Pawn.BLACK && board.get(Coords.D6) == Pawn.BLACK)
                    board.remove(bottom);
            } else if (bottom.equals(Coords.F5)) {
                if (board.get(Coords.G5) == Pawn.BLACK && board.get(Coords.F6) == Pawn.BLACK)
                    board.remove(bottom);
            }

            //sono fuori dalle zone del trono
            else {
                Coords secondPawn = new Coords(toRow + 2, toColumn);
                if (board.get(secondPawn) == Pawn.BLACK || board.inCitadels(secondPawn))
                    board.remove(bottom);
            }
        }

        //capture on top
        else if (toRow > 1 && board.get(top) == Pawn.KING) {
            //re sul trono
            if (top.equals(Coords.E5)) {
                if (board.get(Coords.E4) == Pawn.BLACK
                        && board.get(Coords.F5) == Pawn.BLACK
                        && board.get(Coords.D5) == Pawn.BLACK)
                    board.remove(top);
            }

            //re adiacente al trono
            else if (top.equals(Coords.E6)) {
                if (board.get(Coords.D6) == Pawn.BLACK && board.get(Coords.F6) == Pawn.BLACK)
                    board.remove(top);
            } else if (top.equals(Coords.D5)) {
                if (board.get(Coords.C5) == Pawn.BLACK && board.get(Coords.D4) == Pawn.BLACK)
                    board.remove(top);
            } else if (top.equals(Coords.F5)) {
                if (board.get(Coords.G5) == Pawn.BLACK && board.get(Coords.F4) == Pawn.BLACK)
                    board.remove(top);
            }

            //sono fuori dalle zone del trono
            else {
                Coords secondPawn = new Coords(toRow - 2, toColumn);
                if (board.get(secondPawn) == Pawn.BLACK || board.inCitadels(secondPawn))
                    board.remove(top);
            }
        }
    }
}
